README.md
<!------------------------------------------------------------------------------
--
-- Copyright (C) 2022 Kevin Matthes
--
-- This program is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 2 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along
-- with this program; if not, write to the Free Software Foundation, Inc.,
-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
--
----
--
--  FILE
--      README.md
--
--  BRIEF
--      Important information regarding this project.
--
--  AUTHOR
--      Kevin Matthes
--
--  COPYRIGHT
--      (C) 2022 Kevin Matthes.
--      This file is licensed GPL 2 as of June 1991.
--
--  DATE
--      2022
--
--  NOTE
--      See `LICENSE' for full license.
--
------------------------------------------------------------------------------->

# libgpl2

## Summary

A set of utilities concerning the GPL 2.0 license.

## License

This project's license is **GPL 2** (as of June 1991).  The whole license text
can be found in `LICENSE` in the main directory of this repository.  The brief
version is as follows:

> Copyright (C) 2022 Kevin Matthes
>
> This program is free software; you can redistribute it and/or modify
> it under the terms of the GNU General Public License as published by
> the Free Software Foundation; either version 2 of the License, or
> (at your option) any later version.
>
> This program is distributed in the hope that it will be useful,
> but WITHOUT ANY WARRANTY; without even the implied warranty of
> MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
> GNU General Public License for more details.
>
> You should have received a copy of the GNU General Public License along
> with this program; if not, write to the Free Software Foundation, Inc.,
> 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

When compiling a printable version of this documentation using Pandoc, the full
license will be attached automatically to the resulting document.  This can be
invoked by calling `repository-manual.m`.

## Software Requirements

| Requirement       | Type          | Role                                  |
|:------------------|:-------------:|:--------------------------------------|
| `ar`              | application   | create and update static libraries    |
| `docs-snippets`   | repository    | documentation constants               |
| Doxygen           | application   | creation of source code documentation |
| GCC               | application   | C compiler                            |
| GNU Octave        | application   | execution of the provided scripts     |
| `make`            | application   | finalisation of Doxygen documentation |
| Pandoc            | application   | compilation of repository manual      |
| `texlive-full`    | package       | compilation of repository manual      |

The compilation of such an **optional** repository manual can be invoked by just
calling one of the following lines in a terminal.

```
octave repository-manual.m
octave-cli repository-manual.m
```

Both will redirect to the same application, GNU Octave, which will then create
the manual for this repository and attach the entire license to it.  The
resulting file, `repository.pdf`, will be saved in the main directory of this
repository.

<!----------------------------------------------------------------------------->


CHANGELOG.md
<!------------------------------------------------------------------------------
--
-- Copyright (C) 2022 Kevin Matthes
--
-- This program is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 2 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along
-- with this program; if not, write to the Free Software Foundation, Inc.,
-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
--
----
--
--  FILE
--      CHANGELOG.md
--
--  BRIEF
--      The development history of this project.
--
--  AUTHOR
--      Kevin Matthes
--
--  COPYRIGHT
--      (C) 2022 Kevin Matthes.
--      This file is licensed GPL 2 as of June 1991.
--
--  DATE
--      2022
--
--  NOTE
--      See `LICENSE' for full license.
--      See `README.md' for project details.
--
------------------------------------------------------------------------------->

# Changelog

## [Unreleased]

## [1.2] -- 2022-05-16

### Added

* submodule `blank`
* submodule `doxygen.cfg`

### Changed

* optimise repository manual build instruction
* unite `authors.yaml` and `project.yaml`
* update Doxygen settings to version 1.9.1
* update `docs-snippets`
* update `.gitignore`
* update `routines`

### Fixed

* incorrect `CITATION.cff`

### Removed

* `authors.yaml`

## [1.1] -- 2022-04-27

### Added

* create `CITATION.cff`

### Changed

* shorten project description

## [1.0] -- 2022-04-26

### Added

* build instruction:  Doxygen documentation
* Doxygen settings

### Changed

* adjust README description
* update `docs-snippets`

### Fixed

* add missing docstring
* add missing changelog items
* wrong docstrings

## [0.2] -- 2022-04-24

### Added

* build instruction:  library creation
* create `gcc-ar.m`
* import `GPL_2.0.h` from `italiano`
* import `gpl2_copyright_line.c` from `italiano`
* import `gpl2_information.c` from `italiano`
* import `gpl2_license.c` from `italiano`
* import `gpl2_version.c` from `italiano`
* import `gpl2_warranty.c` from `italiano`

### Removed

* submodule `italiano`

## [0.1] -- 2022-04-24

### Added

* add GPL 2.0 license
* build instruction:  repository manual
* create `.gitignore`
* create list of authors
* create project meta data for repository manual
* create repository README
* create this changelog
* create this repository
* submodule `docs-snippets`
* submodule `italiano`
* submodule `routines`

### Deprecated

* submodule `italiano`

<!----------------------------------------------------------------------------->


GPL_2.0.h
/*
 * Copyright (C) 2022 Kevin Matthes
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/**
 * \author      Kevin Matthes
 * \brief       The brief license text of the GPL 2.0 license.
 * \copyright   (C) 2022 Kevin Matthes.
 *              This file is licensed GPL 2 as of June 1991.
 * \date        2022
 * \file        GPL_2.0.h
 * \note        See `LICENSE' for full license.
 *              See `README.md' for project details.
 *
 * This header defines utility functions and macros concering the GPL 2.0
 * license.
 */

/******************************************************************************/

/**
 * \brief   Prevent this header from being included twice.
 *
 * In case this header file should be included more than just once, unexpected
 * side effects might take place.  This unintended behaviour will be avoided by
 * the definition of this macro.
 */

#pragma once
#ifndef __GPL_2_0_H__
#define __GPL_2_0_H__



/*
 * Includes.
 */

// The C standard library.
#include <stdbool.h>
#include <stdio.h>



/**
 * \brief   The GPL 2.0 license (brief version).
 *
 * This macro represents the brief version of the GPL 2.0 license text.
 */

#define GPL_2_0                                                                \
    "This program is free software; you can redistribute it and/or modify\n"   \
    "it under the terms of the GNU General Public License as published by\n"   \
    "the Free Software Foundation; either version 2 of the License, or\n"      \
    "(at your option) any later version.\n"                                    \
    "\n"                                                                       \
    "This program is distributed in the hope that it will be useful,\n"        \
    "but WITHOUT ANY WARRANTY; without even the implied warranty of\n"         \
    "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n"          \
    "GNU General Public License for more details.\n"                           \
    "\n"                                                                       \
    "You should have received a copy of the GNU General Public License along\n"\
    "with this program; if not, write to the Free Software Foundation, Inc.,\n"\
    "51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.\n"



/**
 * \brief   The GPL 2.0 warranty.
 *
 * This macro represents the entire warranty section of the GPL 2.0 license.
 */

#define GPL_2_0__WARRANTY                                                       \
"                            NO WARRANTY\n"                                     \
"\n"                                                                            \
"  11. BECAUSE THE PROGRAM IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY\n"  \
"FOR THE PROGRAM, TO THE EXTENT PERMITTED BY APPLICABLE LAW.  EXCEPT WHEN\n"    \
"OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES\n"      \
"PROVIDE THE PROGRAM \"AS IS\" WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED\n"\
"OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF\n"        \
"MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.  THE ENTIRE RISK AS\n"   \
"TO THE QUALITY AND PERFORMANCE OF THE PROGRAM IS WITH YOU.  SHOULD THE\n"      \
"PROGRAM PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL NECESSARY SERVICING,\n"    \
"REPAIR OR CORRECTION.\n"                                                       \
"\n"                                                                            \
"  12. IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING\n" \
"WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR\n"         \
"REDISTRIBUTE THE PROGRAM AS PERMITTED ABOVE, BE LIABLE TO YOU FOR DAMAGES,\n"  \
"INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES ARISING\n" \
"OUT OF THE USE OR INABILITY TO USE THE PROGRAM (INCLUDING BUT NOT LIMITED\n"   \
"TO LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY\n"    \
"YOU OR THIRD PARTIES OR A FAILURE OF THE PROGRAM TO OPERATE WITH ANY OTHER\n"  \
"PROGRAMS), EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE\n"       \
"POSSIBILITY OF SUCH DAMAGES.\n"



/*
 * Function declarations.
 */

extern  bool    gpl2_copyright_line ( FILE * const          stream
                                    , const int             year_first
                                    , const int             year_latest
                                    , const char * const    people
                                    );
extern  bool    gpl2_information    ( FILE * const          stream
                                    , const char * const    app_name
                                    , const int             major
                                    , const int             minor
                                    , const int             fix_level
                                    , const int             year_first
                                    , const int             year_latest
                                    , const char * const    people
                                    );
extern  bool    gpl2_license        ( FILE * const          stream
                                    , const int             year_first
                                    , const int             year_latest
                                    , const char * const    people
                                    );
extern  bool    gpl2_version        ( FILE * const          stream
                                    , const char * const    app_name
                                    , const int             major
                                    , const int             minor
                                    , const int             fix_level
                                    );
extern  bool    gpl2_warranty       (FILE * const stream);



/*
 * End of header.
 */

// Leaving the header.
#endif  // ! __GPL_2_0_H__

/******************************************************************************/


doxygen.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% Copyright (C) 2022 Kevin Matthes
%%
%% This program is free software; you can redistribute it and/or modify
%% it under the terms of the GNU General Public License as published by
%% the Free Software Foundation; either version 2 of the License, or
%% (at your option) any later version.
%%
%% This program is distributed in the hope that it will be useful,
%% but WITHOUT ANY WARRANTY; without even the implied warranty of
%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%% GNU General Public License for more details.
%%
%% You should have received a copy of the GNU General Public License along
%% with this program; if not, write to the Free Software Foundation, Inc.,
%% 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
%%
%%%%
%%
%%  FILE
%%      doxygen.m
%%
%%  BRIEF
%%      Create a complete documentation from sources using `doxygen`.
%%
%%  AUTHOR
%%      Kevin Matthes
%%
%%  COPYRIGHT
%%      (C) 2022 Kevin Matthes.
%%      This file is licensed GPL 2 as of June 1991.
%%
%%  DATE
%%      2022
%%
%%  NOTE
%%      See `LICENSE' for full license.
%%      See `README.md' for project details.
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%
%%
%% Variables.
%%
%%%%

% Software.
software.compiler.self  = ' doxygen ';

software.make.self  = ' make ';
software.make.flags = ' -C ';
software.make.call  = [software.make.self software.make.flags];



% Directories.
directories.doxygen.make    = './latex/';



% Files.
files.refman.pdf    = [directories.doxygen.make 'refman.pdf'];
files.refman.tex    = [directories.doxygen.make 'refman.tex'];

files.self      = 'doxygen.m';
files.source    = 'libgpl2.doxygen.cfg';
files.target    = 'libgpl2.pdf';



% Control flow.
banner  = ['[ ' files.self ' ] '];



% Call adjustment.
software.compiler.call  = [software.compiler.self files.source];

software.make.call  = [software.make.call directories.doxygen.make];



%%%%
%%
%% Build steps.
%%
%%%%

% Begin build instruction.
disp ([banner 'Begin build instruction.']);



% Call Doxygen.
disp ([banner 'Compile Doxygen documentation ...']);

disp (software.compiler.call);
software.compiler.result = system (software.compiler.call);

disp ([banner 'Done.']);



% Compile LaTeX manual, if possible.
if ~ software.compiler.result && length (glob (files.refman.tex)) && isunix;
    disp ([banner 'Compile LaTeX manual ...']);

    disp (software.make.call);
    unix (software.make.call);

    if length (glob (files.refman.pdf));
        movefile (files.refman.pdf, files.target, 'f');
    end;

    disp ([banner 'Done.']);
end;



% End build instruction.
disp ([banner 'End build instruction.']);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


gpl2_license.c
/*
 * Copyright (C) 2022 Kevin Matthes
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/**
 * \author      Kevin Matthes
 * \brief       The GPL 2.0 license header.
 * \copyright   (C) 2022 Kevin Matthes.
 *              This file is licensed GPL 2 as of June 1991.
 * \date        2022
 * \file        gpl2_license.c
 * \note        See `LICENSE' for full license.
 *              See `README.md' for project details.
 *
 * This file implements the GPL 2.0 license header function.  The text will be
 * written to the given stream.
 */

/******************************************************************************/

/*
 * Includes.
 */

// This project.
#include "GPL_2.0.h"



/**
 * \brief   Print the copyright line of the GPL 2.0 license header.
 * \param   stream      The stream to write the copyright line to.
 * \param   year_first  The year of the first commit.
 * \param   year_latest The year of the latest commit.
 * \param   people      The people who contributed.
 * \return  Whether the text was printed successfully.
 *
 * Write the GPL 2.0 license header's copyright line to the given stream.
 */

inline bool gpl2_license ( FILE * const          stream
                         , const int             year_first
                         , const int             year_latest
                         , const char * const    people
                         )
{
    return gpl2_copyright_line (stream, year_first, year_latest, people)
        &&             fprintf (stream, GPL_2_0)
         ;
}

/******************************************************************************/


gpl2_version.c
/*
 * Copyright (C) 2022 Kevin Matthes
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/**
 * \author      Kevin Matthes
 * \brief       A minimum version information line.
 * \copyright   (C) 2022 Kevin Matthes.
 *              This file is licensed GPL 2 as of June 1991.
 * \date        2022
 * \file        gpl2_version.c
 * \note        See `LICENSE' for full license.
 *              See `README.md' for project details.
 *
 * This file implements a minimal version information function.  The text will
 * be written to the given stream using **hexadecimal integers**.
 */

/******************************************************************************/

/*
 * Includes.
 */

// This project.
#include "GPL_2.0.h"



/**
 * \brief   A minimum version information.
 * \param   stream      The stream to write the information to.
 * \param   app_name    The concerning application's name.
 * \param   major       The concerning application's major version.
 * \param   minor       The concerning application's minor version.
 * \param   fix_level   The concerning application's fix level.
 * \return  Whether the text was printed successfully.
 *
 * This function will write a brief version information for the concerning
 * application to the given stream.  Please note that the version will be output
 * using **hexadecimal integer** numbers.
 */

inline bool gpl2_version ( FILE * const          stream
                         , const char * const    app_name
                         , const int             major
                         , const int             minor
                         , const int             fix_level
                         )
{
    return fix_level < 0x0 ? fprintf ( stream
                                     , "%s, version %x.%x.\n"
                                     , app_name
                                     , major
                                     , minor
                                     )
                           : fprintf (stream
                                     , "%s, version %x.%x.%x.\n"
                                     , app_name
                                     , major
                                     , minor
                                     , fix_level
                                     )
                           ;
}

/******************************************************************************/


gpl2_warranty.c
/*
 * Copyright (C) 2022 Kevin Matthes
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/**
 * \author      Kevin Matthes
 * \brief       The GPL 2.0 warranty information.
 * \copyright   (C) 2022 Kevin Matthes.
 *              This file is licensed GPL 2 as of June 1991.
 * \date        2022
 * \file        gpl2_warranty.c
 * \note        See `LICENSE' for full license.
 *              See `README.md' for project details.
 *
 * This file implements the GPL 2.0 warranty function.  It will write a brief
 * warranty notice to the given stream and than cite the according section of
 * the GPL 2.0 license.
 */

/******************************************************************************/

/*
 * Includes.
 */

// This project.
#include "GPL_2.0.h"



/**
 * \brief   Write the GPL 2.0 warranty information to the given stream.
 * \param   stream  The stream to write the GPL 2.0 warranty information to.
 * \return  Whether the text was printed successfully.
 *
 * This function will write a warranty information to the given stream.
 */

inline bool gpl2_warranty (FILE * const stream)
{
    return fprintf ( stream
                   , "This application comes with ABSOLUTELY NO WARRANTY, not"
                     " even for MERCHANTABILITY\nor FITNESS FOR A PARTICULAR "
                     "PURPOSE.\n\nThe GPL 2.0 license specifies the following"
                     ":\n\n" GPL_2_0__WARRANTY
                   );
}

/******************************************************************************/


gpl2_information.c
/*
 * Copyright (C) 2022 Kevin Matthes
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/**
 * \author      Kevin Matthes
 * \brief       The GPL 2.0 information.
 * \copyright   (C) 2022 Kevin Matthes.
 *              This file is licensed GPL 2 as of June 1991.
 * \date        2022
 * \file        gpl2_information.c
 * \note        See `LICENSE' for full license.
 *              See `README.md' for project details.
 *
 * This file implements the GPL 2.0 information function.  The text will be
 * written to the given stream.
 */

/******************************************************************************/

/*
 * Includes.
 */

// This project.
#include "GPL_2.0.h"



/**
 * \brief   The information for the interactive mode.
 * \param   stream      The stream to write the information to.
 * \param   app_name    The concerning application's name.
 * \param   major       The concerning application's major version.
 * \param   minor       The concerning application's minor version.
 * \param   fix_level   The concerning application's fix level.
 * \param   year_first  The year of the first commit.
 * \param   year_latest The year of the latest commit.
 * \param   people      The people who contributed.
 * \return  Whether the text was printed successfully.
 *
 * This function will write a brief information block to the given stream.  This
 * is useful especially for the interactive mode.
 */

inline bool gpl2_information ( FILE * const          stream
                             , const char * const    app_name
                             , const int             major
                             , const int             minor
                             , const int             fix_level
                             , const int             year_first
                             , const int             year_latest
                             , const char * const    people
                             )
{
    return        gpl2_version ( stream, app_name,   major, minor, fix_level )
        && gpl2_copyright_line ( stream, year_first, year_latest,  people    )
        &&             fprintf ( stream
                               , "This application is free software.  For copyi"
                                 "ng details, see\nthe GNU General Public Licen"
                                 "se, version 2, as of June 1991.\n\nThis appli"
                                 "cation comes with ABSOLUTELY NO WARRANTY, not"
                                 " even\nfor MERCHANTABILITY or FITNESS FOR A P"
                                 "ARTICULAR PURPOSE;\nrun `%s -w' for more deta"
                                 "ils.\n\n"
                               , app_name
                               )
         ;
}

/******************************************************************************/


gpl2_copyright_line.c
/*
 * Copyright (C) 2022 Kevin Matthes
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/**
 * \author      Kevin Matthes
 * \brief       The copyright line of the GPL 2.0 license header.
 * \copyright   (C) 2022 Kevin Matthes.
 *              This file is licensed GPL 2 as of June 1991.
 * \date        2022
 * \file        gpl2_copyright_line.c
 * \note        See `LICENSE' for full license.
 *              See `README.md' for project details.
 *
 * This file implements the copyright line function of the GPL 2.0 license
 * header.  It will be written to the given stream.
 */

/******************************************************************************/

/*
 * Includes.
 */

// This project.
#include "GPL_2.0.h"



/**
 * \brief   Print the copyright line of the GPL 2.0 license header.
 * \param   stream      The stream to write the copyright line to.
 * \param   year_first  The year of the first commit.
 * \param   year_latest The year of the latest commit.
 * \param   people      The people who contributed.
 * \return  Whether the text was printed successfully.
 *
 * Write the GPL 2.0 license header's copyright line to the given stream.
 */

inline bool gpl2_copyright_line ( FILE * const          stream
                                , const int             year_first
                                , const int             year_latest
                                , const char * const    people
                                )
{
    return year_first >= year_latest ? fprintf ( stream
                                               , "Copyright (C) %d %s\n\n"
                                               , year_first
                                               , people
                                               )
                                     : fprintf ( stream
                                               , "Copyright (C) %d─%d %s\n\n"
                                               , year_first
                                               , year_latest
                                               , people
                                               )
                                     ;
}

/******************************************************************************/


gcc-ar.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% Copyright (C) 2022 Kevin Matthes
%%
%% This program is free software; you can redistribute it and/or modify
%% it under the terms of the GNU General Public License as published by
%% the Free Software Foundation; either version 2 of the License, or
%% (at your option) any later version.
%%
%% This program is distributed in the hope that it will be useful,
%% but WITHOUT ANY WARRANTY; without even the implied warranty of
%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%% GNU General Public License for more details.
%%
%% You should have received a copy of the GNU General Public License along
%% with this program; if not, write to the Free Software Foundation, Inc.,
%% 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
%%
%%%%
%%
%%  FILE
%%      gcc-ar.m
%%
%%  BRIEF
%%      Create a static library from C source code using GCC and `ar`.
%%
%%  AUTHOR
%%      Kevin Matthes
%%
%%  COPYRIGHT
%%      (C) 2022 Kevin Matthes.
%%      This file is licensed GPL 2 as of June 1991.
%%
%%  DATE
%%      2022
%%
%%  NOTE
%%      See `LICENSE' for full license.
%%      See `README.md' for project details.
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%
%%
%% Variables.
%%
%%%%

% Software.
archiver.args   = 'rsv';
archiver.in     = '*.o';
archiver.out    = ['lib' 'gpl2' '.a'];
archiver.self   = 'ar';
archiver.call   = [ archiver.self ' ' archiver.args ' ' archiver.out ' '     ...
                    archiver.in                                              ...
                  ];

compiler.args   = ' -Wall -Werror -Wextra -Wpedantic -std=c99 -c ';
compiler.in     = '*.c';
compiler.out    = '*.o';
compiler.self   = 'gcc';
compiler.call   = [compiler.self ' ' compiler.args ' ' compiler.in];



% Miscellaneous.
misc.self   = 'gcc-ar.m';
misc.banner = ['[ ' misc.self ' ] '];



%%%%
%%
%% Build steps.
%%
%%%%

% Begin build instruction.
disp ([misc.banner 'Begin build instruction.']);



% Clean outdated library.
fprintf ([misc.banner 'Remove outdated version of ' archiver.out ' ... ']);

if length (glob (archiver.out));
    delete (archiver.out);
end;

disp ('Done.');



% Call C compiler.
disp ([misc.banner 'Compile object files ...']);

disp (compiler.call);
system (compiler.call);

disp ([misc.banner 'Done.']);



% Call library creation tool.
if length (glob (archiver.in));
    disp ([misc.banner 'Create library ' archiver.out ' ...']);

    disp (archiver.call);
    system (archiver.call);

    disp ([misc.banner 'Done.']);
end;



% Clean artifacts.
fprintf ([misc.banner 'Remove build artifacts ... ']);

if length (glob (compiler.out));
    delete (compiler.out);
end;

disp ('Done.');



% End build instruction.
disp ([misc.banner 'End build instruction.']);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
